drivers/samsung/displays/mdc_protocol.cr

```diff
@@ -0,0 +1,364 @@
+        data = {{name.id.stringify}} == "auto_off_timer" ? Bytes[0x81, state] : state
+      {% elsif kind.resolve == Int32 %}
+        data = value.clamp(0, 100)
+      {% end %}
+      do_send(Command.parse({{name.id.stringify}}), data, **options)
+    end
+  {% end %}
+
+  def do_device_config
+    {% for name, kind in DEVICE_SETTINGS %}
+      %value = setting?({{kind}}, {{name.id.stringify}})
+      {{name.id}}(%value) unless %value.nil?
+    {% end %}
+  end
+
+  enum ResponseStatus
+    Ack = 0x41 # A
+    Nak = 0x4e # N
+  end
+
+  def received(data, task)
+    hex = data.hexstring
+    logger.debug { "Samsung sent: #{hex}" }
+    data = data.map(&.to_i).to_a
+
+    # Calculate checksum of response
+    checksum = data[1..-2].reduce(&.+)
+
+    # Pop also removes the checksum from the response here
+    if data.pop != checksum
+      logger.error { "Invalid checksum\nChecksum should be: #{checksum.to_s(16)}" }
+      return task.try &.retry
+    end
+
+    status = ResponseStatus.from_value(data[4])
+    command = Command.from_value(data[5])
+    values = data[6..-1]
+    value = values.first
+
+    case status
+    when .ack?
+      case command
+      when .status?
+        self[:hard_off]   = hard_off = values[0] == 0
+        self[:power]      = false if hard_off
+        self[:volume]     = values[1]
+        self[:audio_mute] = values[2] == 1
+        self[:input]      = Inputs.from_value(values[3]).to_s
+        check_power_state
+      when .panel_mute?
+        self[:power] = value == 0
+        check_power_state
+      when .volume?
+        self[:volume] = value
+        self[:audio_mute] = false if value > 0
+      when .brightness?
+        self[:brightness] = value
+      when .input?
+        self[:input] = Inputs.from_value(value).to_s
+        # The input feedback behaviour seems to go a little odd when
+        # screen split is active. Ignore any input forcing when on.
+        unless (screen_split = self[:screen_split]?) && screen_split.as_bool?
+          input_target = @input_target
+          @input_stable = self[:input]? == input_target
+          if input_target
+            switch_to(input_target) unless @input_stable
+          end
+        end
+      when .speaker?
+        self[:speaker] = SpeakerModes.from_value(value).to_s
+      when .hard_off?
+        unless (was_off = self[:hard_off]?) && was_off.as_bool?
+          self[:hard_off] = hard_off = value == 0
+          self[:power] = false if hard_off
+        end
+      when .screen_split?
+        self[:screen_split] = value >= 0
+      when .software_version?
+        self[:software_version] = values.join
+      when .serial_number?
+        self[:serial_number] = values.join
+      else
+        logger.debug { "Samsung responded with ACK: #{value}" }
+      end
+
+      task.try &.success
+    when .nak?
+      task.try &.abort("Samsung responded with NAK: #{hex}")
+    else
+      task.try &.abort("Samsung aborted with: #{hex}")
```

Comment:
If the response is neither a nak or ack, it may be worth retrying to task instead of aborting.